lib/THCUNN/SparseLinear.cu

```diff
@@ -0,0 +1,284 @@
+      THCudaIntTensor_data(state, csr_int),
+      THCudaIntTensor_data(state, colinds_int),
+      THCudaTensor_data(state, weight), inDim,
+      &one, THCudaTensor_data(state, output), batchnum
+  );
+  cusparseDestroyMatDescr(descr); 
+  descr = 0;
+  THCudaTensor_free(state, colinds);
+  THCudaTensor_free(state, values_sel);
+  THCudaTensor_free(state, values);
+  THCudaIntTensor_free(state, colinds_int);
+  THCudaIntTensor_free(state, csr_int);
+#endif
+  THCudaTensor_free(state, output_row);
+}
+
+void THNN_CudaSparseLinear_zeroGradParameters(
+          THCState *state,
+          THCudaTensor *gradWeight,
+          THCudaTensor *gradBias,
+          THCudaTensor *lastInput)
+{
+  long h, i;
+  long outDim = THCudaTensor_size(state, gradWeight, 0);
+  long inDim = THCudaTensor_size(state, gradWeight, 1);
+  THCudaTensor *tmp = THCudaTensor_new(state);
+
+  THArgCheck(checkSize1D(gradBias, outDim), 3, "gradBias size wrong");
+  THArgCheck(checkInput(lastInput), 4,
+      "input size must be batchsize x nnz x 2");
+  THArgCheck(THCudaTensor_isContiguous(state, gradWeight), 2,
+      "gradWeight must be contiguous to use cuda");
+
+  if (!THCudaTensor_isContiguous(state, gradWeight))
+    gradWeight = THCudaTensor_newContiguous(state, gradWeight);
+  THCudaTensor_zero(state, gradBias);
+
+  long batchnum = THCudaTensor_size(state, lastInput, 0);
+  long nnz = THCudaTensor_size(state, lastInput, 1);
+
+  for (h = 0; h < batchnum; h++) {
+    for (i = 0; i < nnz; i++)
+    {
+      long offset = (long)(THCudaTensor_get3d(state, lastInput, h, i, 0)) - 1;
+
+      if (offset >= 0 && offset < inDim) {
+        THCudaTensor_select(state, tmp, gradWeight, 1, offset);
+        THCudaTensor_zero(state, tmp);
+      } else {
+        THError("index out of bound. zeroGradParameters: \
+            %ld not between 1 and %ld", offset + 1, inDim);
+      }
+    }
+  }
+}
+
+void THNN_CudaSparseLinear_accGradParameters(
+    THCState *state,
+    THCudaTensor *input,
+    THCudaTensor *gradOutput,
+    THCudaTensor *gradWeight,
+    THCudaTensor *gradBias,
+    THCudaTensor *weight,
+    THCudaTensor *bias,
+    double weightDecay,
+    double scale)
+{
+  long i, h;
+  long outDim = THCudaTensor_size(state, weight, 0);
+  long inDim = THCudaTensor_size(state, weight, 1);
+
+  THArgCheck(checkInput(input), 2, "input size must be batchsize x nnz x 2");
+  THArgCheck(checkSize2D(gradWeight, outDim, inDim), 4, "gradWeight size wrong");
+  THArgCheck(checkSize1D(gradBias, outDim), 5, "gradBias size wrong");
+  THArgCheck(THCudaTensor_isContiguous(state, gradOutput), 3,
+             "gradOutput must be contiguous");
+
+  long batchnum = THCudaTensor_size(state, input, 0);
+  long nnz = THCudaTensor_size(state, input, 1);
+  THCudaTensor_resize2d(state, gradOutput, batchnum, outDim);
+
+  for (h = 0; h < batchnum; h++) {
+    for (i = 0; i < nnz; i++) {
+      float val = scale*THCudaTensor_get3d(state, input, h, i, 1);
```

Comment:
this is a blocking operation as well. 
let's discuss this PR offline. a bit more work has to be done to remove bottlenecks.